internal/mode/static/policies/clientsettings/validator.go

```diff
@@ -0,0 +1,215 @@
+			fmt.Sprintf("unsupported targetRef Group %q; must be %s", ref.Group, gatewayv1.GroupName),
+		)
+	}
+
+	supportedKinds := []gatewayv1.Kind{kinds.Gateway, kinds.HTTPRoute, kinds.GRPCRoute}
+
+	if !slices.Contains(supportedKinds, ref.Kind) {
+		path := basePath.Child("kind")
+
+		return field.Invalid(
+			path,
+			ref.Kind,
+			fmt.Sprintf("unsupported targetRef Kind %q; Kind must be one of: %v", ref.Kind, supportedKinds),
+		)
+	}
+
+	return nil
+}
+
+func (v *Validator) validateSettings(spec ngfAPI.ClientSettingsPolicySpec) error {
+	var allErrs field.ErrorList
+	fieldPath := field.NewPath("spec")
+
+	if spec.Body != nil {
+		allErrs = append(allErrs, v.validateClientBody(*spec.Body, fieldPath.Child("body"))...)
+	}
+
+	if spec.KeepAlive != nil {
+		allErrs = append(allErrs, v.validateClientKeepAlive(*spec.KeepAlive, fieldPath.Child("keepAlive"))...)
+	}
+
+	return allErrs.ToAggregate()
+}
+
+func (v *Validator) validateClientBody(body ngfAPI.ClientBody, fieldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if body.Timeout != nil {
+		if err := v.genericValidator.ValidateNginxDuration(string(*body.Timeout)); err != nil {
+			path := fieldPath.Child("timeout")
+
+			allErrs = append(allErrs, field.Invalid(path, body.Timeout, err.Error()))
+		}
+	}
+
+	if body.MaxSize != nil {
+		if err := v.genericValidator.ValidateNginxSize(string(*body.MaxSize)); err != nil {
+			path := fieldPath.Child("maxSize")
+
+			allErrs = append(allErrs, field.Invalid(path, body.MaxSize, err.Error()))
+		}
+	}
+
+	return allErrs
+}
+
+func (v *Validator) validateClientKeepAlive(keepAlive ngfAPI.ClientKeepAlive, fieldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if keepAlive.Requests != nil {
+		requests := *keepAlive.Requests
+		if requests < 0 {
+			path := fieldPath.Child("requests")
+
+			allErrs = append(
+				allErrs,
+				field.Invalid(path, *keepAlive.Requests, "requests is invalid: must be positive"),
+			)
+		}
+	}
+
+	if keepAlive.Time != nil {
+		if err := v.genericValidator.ValidateNginxDuration(string(*keepAlive.Time)); err != nil {
+			path := fieldPath.Child("time")
+
+			allErrs = append(allErrs, field.Invalid(path, *keepAlive.Time, err.Error()))
+		}
+	}
+
+	if keepAlive.Timeout != nil {
+		timeout := keepAlive.Timeout
+
+		if timeout.Server != nil {
+			if err := v.genericValidator.ValidateNginxDuration(string(*timeout.Server)); err != nil {
+				path := fieldPath.Child("timeout").Child("server")
+
+				allErrs = append(
+					allErrs,
+					field.Invalid(path, *keepAlive.Timeout.Server, err.Error()),
+				)
+			}
+		}
+
+		if timeout.Header != nil {
+			if err := v.genericValidator.ValidateNginxDuration(string(*timeout.Header)); err != nil {
+				path := fieldPath.Child("timeout").Child("header")
+
+				allErrs = append(
+					allErrs,
+					field.Invalid(path, *keepAlive.Timeout.Header, err.Error()),
+				)
+			}
+		}
+
+		if keepAlive.Timeout.Header != nil && keepAlive.Timeout.Server == nil {
```

Comment:
CEL can be bypassed like CRD validation. This is back to the same discussion....

Reply:
Right, but is this a security risk? Or just an invalid config?

Reply:
It's just invalid config. But shouldn't we be trying to prevent that? We don't have a way to report nginx reload errors on a policy or a gateway api resource. This makes it really hard to resolve an invalid config issue. You need access to the nginx container logs and knowledge of the policy.

Reply:
If we decide to do (2), then we should go back and re-implement all of the validation for all Gateway resources that we removed when the Gateway API team removed it. I thought we had consensus at that point that we didn't want to re-validate, but maybe that opinion has changed.

IMO, it's overkill and does require work, testing, and maintenance. We would also always need to be in sync with the various schemas, which is a risk. CEL and the openapi schemas were introduced so implementers don't have to do this.

I understand the benefits of multiple lines of defense, which is why I'm happy with double validation for security-related fields, but at some point it feels like a slippery slope.

Reply:
Now, per my other comment about embedding the schema in our code, then we don't have to worry about the issues I mention. But that will require some work.

Reply:
I'm ok with option #1 for now, and filing a ticket to investigate embedding the schema in our code.